File: MCP/MCP.go
package main

import
(
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
	"math/rand"
)

const NUM_THREAD=4;

var count []int
var tid [NUM_THREAD]struct{}
var totalPoint int

//function for goroutines; estimating pi
func circlePoint(countIndex int) {
	localCount := 0
	//for each point
	for i := int(0); i<totalPoint; i++{
		//get random coordinates
		x := rand.Float64()
		y := rand.Float64()
		//determine if point falls in circle
		r := x*x + y*x
		//if point is in circle, increase localCount
		if r <= 1 {
			localCount++
		}
	}
	//add local count to array
	count[countIndex] = localCount
}

func main() {
	//error checking
	if len(os.Args) != 2{
		fmt.Println("Argument is wrong")
		return
	}


	arg, err := strconv.Atoi(os.Args[1])


	if err!=nil || arg<=0{
		fmt.Println("Invalid argument")
		return
	}

	//how many points each goroutine will calculate
	totalPoint = arg/NUM_THREAD
	//create a count slice of length NUM_THREAD
	count = make([]int, NUM_THREAD)

	rand.Seed(time.Now().UnixNano())

	//create a wait group
	var wg sync.WaitGroup
	//start clock
	start := time.Now()
	//create goroutines and add them to the waitgroup
	for i :=0; i<NUM_THREAD; i++{
		wg.Add(1)
		go func(index int){
			defer wg.Done()
			circlePoint(index)
		}(i)
	}
	//wait for the goroutines to finish
	wg.Wait()

	//calculate total points that fall in the circle with data from goroutines
	countCircle := 0
	for _, c := range count{
		countCircle += c
	}

	//calculate pi
	pi := 4.0 * float64(countCircle)/float64(totalPoint*NUM_THREAD)
	//to get rid of error
	pi = pi-0

	//end time
	elapsed := time.Since(start)

	//fmt.Printf("PI= %.15f\n", pi)
	fmt.Println(elapsed.Seconds())
}

File: MCP/MCP.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <pthread.h>

#define NUM_THREAD 4

void* circle_point(void *param);

pthread_t tid[NUM_THREAD]={0};
int count[NUM_THREAD]={0};
struct timespec start_time, end_time;
static long int total_point;
static long int count_circle=0;


int main(int argc, char const *argv[]){

   if(argc!=2){
       printf("Argument is wrong\n");
       return -1;
     }

   total_point=atoll(argv[1])/NUM_THREAD;

   clock_gettime(CLOCK_MONOTONIC, &start_time); 
   srand(time(NULL));
   static int i;
   for(i=0; i<NUM_THREAD;i++)
   	pthread_create(&tid[i],NULL,circle_point,&count[i]);
   for(i=0;i<NUM_THREAD;i++){
       pthread_join(tid[i],NULL);
       count_circle+=count[i];
     }
   double pi=4.0*(double)count_circle/(double)total_point/(double)NUM_THREAD;
   clock_gettime(CLOCK_MONOTONIC, &end_time); 

  // printf("PI = %17.15f\n",pi);
   double elapsed = (end_time.tv_sec - start_time.tv_sec);
   elapsed += (end_time.tv_nsec - start_time.tv_nsec) / 1000000000.0;
   printf("%f", elapsed);
   return 0;
  }

 void* circle_point(void *param){
   int *pcount= (int*)param;
   int i;
   for(i=0; i<total_point;i++){
       double x= (double)rand()/(double)RAND_MAX;
       double y=(double)rand()/(double)RAND_MAX;
       double r= sqrt(x*x+y*y);
       if(r<=1) *pcount=*pcount+1;
     }
   pthread_exit(0);
 }
